src/Native/Runtime/arm/WriteBarriers.S

```diff
@@ -1,68 +1,378 @@
+          // additional memory barrier we don't require but it's on the rare path and x86 doesn't have an xchg
+          // variant that doesn't implicitly specify the lock prefix. Note that INVALIDGCVALUE is a 32-bit
+          // immediate and therefore must be moved into a register before it can be written to the shadow
+          // location.
+          mov          r12, \DESTREG
+          ldr          \DESTREG, [sp]
+          str          r12, [sp]
+          push         \REFREG
+          movw         \REFREG, #(INVALIDGCVALUE & 0xFFFF)
+          movt         \REFREG, #(INVALIDGCVALUE >> 16)
+          str          \REFREG, [\DESTREG]
+          pop          \REFREG
+
+LOCAL_LABEL(\BASENAME\()_UpdateShadowHeap_PopThenDone_\REFREG):
+          // Restore original DESTREG value from the stack.
+          pop          \DESTREG
+
+LOCAL_LABEL(\BASENAME\()_UpdateShadowHeap_Done_\REFREG):
+
+.endm
+
+#else  // WRITE_BARRIER_CHECK
+
+.macro UPDATE_GC_SHADOW BASENAME, REFREG, DESTREG
+.endm
+
+#endif // WRITE_BARRIER_CHECK
+
+// There are several different helpers used depending on which register holds the object reference. Since all
+// the helpers have identical structure we use a macro to define this structure. Two arguments are taken, the
+// name of the register that points to the location to be updated and the name of the register that holds the
+// object reference (this should be in upper case as it's used in the definition of the name of the helper).
+.macro DEFINE_UNCHECKED_WRITE_BARRIER_CORE BASENAME, REFREG
+
+          // Update the shadow copy of the heap with the same value just written to the same heap. (A no-op unless
+          // we're in a debug build and write barrier checking has been enabled).
+          UPDATE_GC_SHADOW \BASENAME, \REFREG, r0
+
+          // If the reference is to an object that's not in an ephemeral generation we have no need to track it
+          // (since the object won't be collected or moved by an ephemeral collection).
+          ldr	         r12, =C_FUNC(g_ephemeral_low)
+          ldr          r12, [r12]
+          cmp          \REFREG, r12
+          blo          LOCAL_LABEL(\BASENAME\()_NoBarrierRequired_\REFREG)
+
+          ldr	         r12, =C_FUNC(g_ephemeral_high)
+          ldr          r12, [r12]
+          cmp	         \REFREG, r12
+          bhi          LOCAL_LABEL(\BASENAME\()_NoBarrierRequired_\REFREG)
+
+          // We have a location on the GC heap being updated with a reference to an ephemeral object so we must
+          // track this write. The location address is translated into an offset in the card table bitmap. We set
+          // an entire byte in the card table since it's quicker than messing around with bitmasks and we only write
+          // the byte if it hasn't already been done since writes are expensive and impact scaling.
+          lsr          r0, #LOG2_CLUMP_SIZE
```

Comment:
This shift should be done as part of address mode - look for add r12, r12, $DESTREG, lsr #10 in the Windows version